### PR TITLE
Terraform

### PR DESCRIPTION
**Pull request by rosemcc**
Adding all Terraform scripts for the Hub stack creation in AWS.

### CODE REVIEW

File: terraform/contentful-backup-bucket.tf [hunk 1]
```diff
@@ -0,0 +1,62 @@
+  ### Maybe this should reside with the serverless.yml for the lambda function? 
+  ### https://github.com/UoA-eResearch/research-hub-contentful-backup/blob/master/serverless.yml
```
Comment (uoa-noel) [inline terraform/contentful-backup-bucket.tf:2]: Might be good to make a decision on this one? The options seem to be keeping the terraform file here, putting the equivalent in the serverless.yml, or perhaps folding the contentful backup repo into the monorepo... :thinking: Your thoughts, Rose?

File: terraform/ecs-service-graphql-preview.tf [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+# cyclic dependency. Means we can also make changes
+# # as needed without impacting the core rules
+# resource "aws_security_group_rule" "lb_to_graphql" {
+#   type                     = "ingress"
+#   from_port                = "0"
+#   to_port                  = "0"
+#   protocol                 = "-1"
+#   source_security_group_id = aws_security_group.loadbalancer_sg.id
+#   security_group_id        = aws_security_group.graphql_sg.id
+# }
```
Comment (uoa-noel) [inline terraform/ecs-service-graphql-preview.tf:203]: Commented out code that can be removed.

File: terraform/ecs-environment.tf [hunk 3]
```diff
@@ -9,7 +9,7 @@ resource "aws_ecr_repository" "graphql" {
   name = var.repository_name
   # Let AWS run vulnerability scanning on image push
   image_scanning_configuration {
-    scan_on_push = true
+    scan_on_push = true   # is false in sandbox... maybe need to check if this breaks anything
```
Comment (uoa-noel) [inline terraform/ecs-environment.tf:12]: Would be good to check if this is working.

**Review by uoa-noel — COMMENTED** [review]
Some well-documented changes to Terraform scripts, just some todos and comments that need to be resolved.

File: terraform/alb-environment.tf [hunk 4]
```diff
@@ -17,7 +17,7 @@ resource "aws_route53_record" "ecs-lb-entry" {
   count   = var.create_dns_entry ? 1 : 0
   zone_id = data.aws_route53_zone.host_zone[count.index].id
   name    = var.lb_dns_name
-  type    = "A"
+  type    = "CNAME"   # TO DO: check if this should be "A"?
```
[the remark this answers is not in the document]
Comment (rosemcc) [reply terraform/alb-environment.tf:20]: Yep, Noel confirmed should be CNAME (based on sandbox config)

File: terraform/cognito.tf [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    scope_description = "Access scope for the Research Hub SPA"
+  }
+  # scope {
+  # scope_name        = "servicenow"
+  #   scope_description = "Access scope for the Research Hub ServiceNow"
+  # }
+  # scope {
+  #   scope_name        = "contentful"
+  #   scope_description = "Access scope for the Research Hub Contentful"
+  # }
```
[the remark this answers is not in the document]
Comment (rosemcc) [reply terraform/cognito.tf:16]: removed

**Comment by rosemcc** [comment]
Need to modify access for the preview env to S3 bucket

File: terraform/ecs-service-graphql-preview.tf (hunk 2 above)
Comment (rosemcc) [reply terraform/ecs-service-graphql-preview.tf:203]: done

File: terraform/ecs-environment.tf (hunk 3 above)
Comment (rosemcc) [reply terraform/ecs-environment.tf:12]: I guess if you were able to deploy to ecr then it's ok..

File: terraform/contentful-backup-bucket.tf (hunk 1 above)
Comment (rosemcc) [reply terraform/contentful-backup-bucket.tf:2]: I think best to migrate the bucket setup to the serverless.yml. That keeps it nicely packaged as a standalone service. 
But we could also then move it into the monorepo - that part is probably a question for the team.

**Pull request merged**